midend/local_copyprop.cpp

```diff
@@ -409,12 +435,21 @@ IR::MethodCallExpression *DoLocalCopyPropagation::postorder(IR::MethodCallExpres
         if (actions.count(fn->path->name)) {
             LOG3("action method call " << mc->method);
             apply_function(&actions[fn->path->name]);
+            return mc;
+        } else if (mi->is<P4::ExternFunction>()) {
+            LOG3("extern function call " << mc->method);
+            // assume it has no side-effects on anything not explicitly passed to it?
+            // maybe should have annotations if it does
```

Comment:
even if it has side-effects it should be unable to change any named variables (locals). We do not have references, so you cannot change something you cannot name - due to copy-in copy-out.

Reply:
it cannot call any methods, since it cannot refer to instances of the allocated objects that have these methods.

Reply:
We don't know for certain that the architecture does not provide a back-door method of referring to other objects -- for example, the packet_in and packet_out objects that are passed to parser and deparsers might be affected by other methods in some way?  Perhaps a clone object that can have a method invoked that triggers an additional copy of the packet?

If we want to allow such architectures, we really need to provide langauge that the architecture can be described in (in the arch.p4 file); I'm not sure annotations are good enough longer-term.  Something for the language design group.